### PR TITLE
Guide to Deploying a Contract from Installed Wasm Bytecode Using Code

### PR DESCRIPTION
**Pull request by KodeSage**
Bounty Submission for #614 

Hello @briwylde08, This is ready for review

### CODE REVIEW

**Comment by briwylde08** [comment]
Thank you for your submission! We'll get it reviewed and let you know any feedback. [🎉](https://emojipedia.org/party-popper)

File: docs/smart-contracts/guides/conventions/deploy-contracts-wasm.mdx [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+git clone -b v20.0.0 https://github.com/stellar/soroban-examples
+```
+
+2. Navigate to the Deployer Example:
+
+```
+cd soroban-examples/deployer/deployer
+```
+
+**NB**: for this example , navigate to `deployer/contracts` and `deployer/deployer` and run the command `stellar contract build` to generate the target files.
```
[the remark this answers is not in the document]
Comment (KodeSage) [reply docs/smart-contracts/guides/conventions/deploy-contracts-wasm.mdx]: alright @briwylde08 , I will change it to `Note` now

**Comment by KodeSage** [comment]
Hi @briwylde08 , I just updated it now.

**Comment by KodeSage** [comment]
@briwylde08 was the main branch updated recently? just realised this is showing a merge conflict.

**Comment by briwylde08** [comment]
Yeah we did a pretty large structural change yesterday that is probably causing the conflict. You can merge the main branch back into your branch, which should help you resolve whatever conflict is triggering the alert.

**Comment by KodeSage** [comment]
> Yeah we did a pretty large structural change yesterday that is probably causing the conflict. You can merge the main branch back into your branch, which should help you resolve whatever conflict is triggering the alert.

Hi @briwylde08 @ElliotFriend , I have Updated this branch with current updated folder structure and resolved the conflict.

**Comment by briwylde08** [comment]
Hi @KodeSage! Can you comment on the corresponding issue here: https://github.com/stellar/stellar-docs/issues/620 so I can assign it to you? Thank you!

**Comment by KodeSage** [comment]
Hi @briwylde08 #614 , this is the Bounty issue

**Comment by briwylde08** [comment]
Yes sorry I was mistaken!

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+  --wasm_hash 6bc6d975ef99c057231481c40f69f4c2a8a89ac56e8826ef0378f8e5c6abc4be \
+  --init_fn init \
+  --init_args '[{"u32":8}]'
+```
+
+- replace `0xkodesage` with your own identity
+
+it returns a Contract identifier `CCTVFX6BFTQHTGAHA5TY4YZQJRUKRE2RRNUTGVBNKE3PJF5C7CI53APY` of the deployed test contract.
+
+Invoke the deployed test contract using the identifier returned from the previous command.
```
Comment (ElliotFriend) [inline docs/build/guides/conventions/deploy-contracts-wasm.mdx:221]:
```suggestion
Invoke the deployed test contract using the address returned from the previous command.
```

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+```sh
+stellar contract invoke \
+  --id CCTVFX6BFTQHTGAHA5TY4YZQJRUKRE2RRNUTGVBNKE3PJF5C7CI53APY \
+  --source 0xkodesage \
+  --network testnet \
+    -- \
+    value
```
Comment (ElliotFriend) [inline docs/build/guides/conventions/deploy-contracts-wasm.mdx:229]:
```suggestion
  -- \
  value
```

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+```
+
+The following output should occur using the code above.
```
Comment (ElliotFriend) [inline docs/build/guides/conventions/deploy-contracts-wasm.mdx:232]:
```suggestion
You should receive something like the following output:
```

**Review by ElliotFriend — COMMENTED** [review]
Looking really good, @KodeSage! I left a couple thoughts, but I think this one is pretty close!

**Comment by KodeSage** [comment]
Hi @ElliotFriend , this changes have been implemented, please do take a look on the updated changes and file.

**Comment by ElliotFriend** [comment]
I'm giving this one another look now, but I can see [it has the extra file](https://github.com/stellar/stellar-docs/compare/main...KodeSage:stellar-docs:deploy_contract_wasm), like in your other PR (#735). Can you get that removed from this PR, too?

**Comment by KodeSage** [comment]
Hi @ElliotFriend , it's all done now. please take a look at it.

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            .deploy(wasm_hash);
+
+        // Invoke the init function with the given arguments.
+        let res: Val = env.invoke_contract(&deployed_address, &init_fn, init_args);
+        // Return the contract ID of the deployed contract and the result of
+        // invoking the init result.
+        (deployed_address, res)
+    }
+}
+
```
Comment (ElliotFriend) [inline docs/build/guides/conventions/deploy-contracts-wasm.mdx:107]:
```suggestion
```

Let's remove this empty line

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+```
+
+#### How it works
```
[the remark this answers is not in the document]
Comment (ElliotFriend) [reply docs/build/guides/conventions/deploy-contracts-wasm.mdx:110]: This is one of the changes that I'd like to see happen

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+The deployer contract provides a mechanism to deploy other contracts in a secure and deterministic manner. It ensures that the deployment is authorized by the specified deployer address and supports atomic initialization of the newly deployed contract. This guarantees that the contract is properly initialized immediately after deployment, preventing any potential issues with uninitialized contracts.
+
+##### **Function breakdown**
```
[the remark this answers is not in the document]
Comment (ElliotFriend) [reply docs/build/guides/conventions/deploy-contracts-wasm.mdx:114]: This is one of the changes that I'd like to see happen

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+- `env: Env`: The Env object represents the current contract execution environment. It provides methods to interact with the blockchain and other contracts, as well as perform various operations.
+- `deployer: Address`: The Address of the entity that is requesting the contract deployment. This address will be used to authorize the deployment.
+- `wasm_hash: BytesN<32>`: The hash of the Wasm bytecode of the contract to be deployed. This hash is used to uniquely identify the contract's code.
+- `salt: BytesN<32>`: A unique value used to derive the address of the deployed contract. The same combination of deployer address and salt will always produce the same deployed contract address, ensuring deterministic contract addresses.
+- `init_fn: Symbol`: The name of the initialization function to be called on the newly deployed contract.
+- `init_args: Vec<Val>`: A vector of arguments to be passed to the initialization function of the deployed contract.
+
+##### **Function steps**
```
[the remark this answers is not in the document]
Comment (ElliotFriend) [reply docs/build/guides/conventions/deploy-contracts-wasm.mdx:123]: This is one of the changes that I'd like to see happen

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+## Run the contract
+
+Before deploying the test contract with the deployer, install the test contract Wasm using the `install` command. The `install` command will print out the hash derived from the Wasm file which should be used by the deployer.
+
+```sh
+stellar contract install --source-account 0xkodesage  --wasm  contract/target/wasm32-unknown-unknown/release/soroban_deployer_test_contract.wasm --rpc-url https://soroban-testnet.stellar.org:443 --network-passphrase "Test SDF Network ; September 2015"   --network testnet
+```
+
+- When you deploy a smart contract to a network, you need to specify an identity that will be used to sign the transactions. Change the `0xkodesage` to your own [identity].
```
[the remark this answers is not in the document]
Comment (ElliotFriend) [reply docs/build/guides/conventions/deploy-contracts-wasm.mdx:186]: This is one of the changes that I'd like to see happen

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+  -- \
+  deploy \
+  --deployer CDKYZMA3OXR54YAHOQ5D4EWDFDT3ZNKKRYKQT7MGPWH22ZVD2DX2BJID \
+  --salt 123 \
+  --wasm_hash 6bc6d975ef99c057231481c40f69f4c2a8a89ac56e8826ef0378f8e5c6abc4be \
+  --init_fn init \
+  --init_args '[{"u32":8}]'
+```
+
+- replace `0xkodesage` with your own identity
```
[the remark this answers is not in the document]
Comment (ElliotFriend) [reply docs/build/guides/conventions/deploy-contracts-wasm.mdx:218]: This is one of the changes that I'd like to see happen

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+
+The deployer contract invocation will return the Contract address (For example: `CCTVFX6BFTQHTGAHA5TY4YZQJRUKRE2RRNUTGVBNKE3PJF5C7CI53APY`) of the newly deployed test contract.
+
+Invoke the deployed test contract using the address returned from the previous command.
+
+```sh
+stellar contract invoke \
+  --id CCTVFX6BFTQHTGAHA5TY4YZQJRUKRE2RRNUTGVBNKE3PJF5C7CI53APY \
+  --source 0xkodesage \
```
[the remark this answers is not in the document]
Comment (ElliotFriend) [reply docs/build/guides/conventions/deploy-contracts-wasm.mdx:227]: This is one of the changes that I'd like to see happen

**Review by ElliotFriend — CHANGES_REQUESTED** [review]
Thanks for making some of the changes, @KodeSage! I've left a few more suggestions on some markdown formatting I missed yesterday. Also, there are still some un-addressed comments/suggested changes from the other review. Not all of them are _must-haves_, of course, but I've noted accordingly any of my comments that are.

Thanks!!

File: docs/build/guides/conventions/deploy-contracts-wasm.mdx
[the remark this answers is not in the document]
Comment (ElliotFriend) [reply docs/build/guides/conventions/deploy-contracts-wasm.mdx:1]: This is one of the changes that I'd like to see happen

**Comment by KodeSage** [comment]
Hi @ElliotFriend , I have implemented the changes. please do take a look at it.

**Comment by KodeSage** [comment]
Hi @ElliotFriend , I have implemented the updated changes. please do take a look at it.

File: docs/build/guides/conventions/deploy-contract.mdx [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+git clone -b v20.2.0 https://github.com/stellar/soroban-examples
+```
+
+2. Navigate to the Deployer Example:
+
+```bash
+cd soroban-examples/deployer/deployer
+```
+
+:::note For this example to work, you should navigate to `deployer/contracts` and `deployer/deployer` and run the command `stellar contract build` in each directory to generate the target files. :::
```
[the remark this answers is not in the document]
Comment (ElliotFriend) [reply docs/build/guides/conventions/deploy-contract.mdx:46]: It probably seems like I'm just nitpicking for no reason here 😅, but this change actually does matter. If there's not a blank line after `:::note`, and another before the closing `:::`, the markdown won't be parsed properly.

**Review by ElliotFriend — CHANGES_REQUESTED** [review]
Hi, @KodeSage! Thanks for making those changes. I have two more nitpicks on the changes for you. It probably seems like I'm just a huge stickler lol. I really appreciate your hard work on this one!

File: docs/build/guides/conventions/deploy-contract.mdx [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+### Function breakdown
+
+- `env: Env`: The Env object represents the current contract execution environment. It provides methods to interact with the blockchain and other contracts, as well as perform various operations.
+- `deployer: Address`: The Address of the entity that is requesting the contract deployment. This address will be used to authorize the deployment.
+- `wasm_hash: BytesN<32>`: The hash of the Wasm bytecode of the contract to be deployed and must already be installed and on the ledger. This hash is used to uniquely identify the contract's code.
+- `salt: BytesN<32>`: A unique value used to derive the address of the deployed contract. The same combination of deployer address and salt will always produce the same deployed contract address, ensuring deterministic contract addresses.
+- `init_fn: Symbol`: The name of the initialization function to be called on the newly deployed contract.
+- `init_args: Vec<Val>`: A vector of arguments, specified as `{type: value}` objects, to be passed to the initialization function of the deployed contract.
+
+##### **Function steps**
```
[the remark this answers is not in the document]
Comment (ElliotFriend) [reply docs/build/guides/conventions/deploy-contract.mdx:121]: I think this one just got missed the first time-round.

**Comment by KodeSage** [comment]
HI @ElliotFriend  , I have updated the changes now. 😁

**Review by ElliotFriend — APPROVED** [review]
Looks great! Thanks for all the hard work!

**Comment by briwylde08** [comment]
Thank you for your work on this @KodeSage! We'll take the conversation back to the corresponding issue as well.

**Pull request merged**